PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,573 @@
+            containing this `ModuleHook` object. If this string is non-unique,
+            an existing in-memory module will be erroneously reused when lazily
+            loading this hook script, thus erroneously resanitizing previously
+            sanitized hook script attributes (e.g., `datas`) with the
+            `format_binaries_and_datas()` helper.
+        """
+
+        # Note that the passed module graph is already a weak reference,
+        # avoiding circular reference issues. See ModuleHookCache.__init__().
+        assert isinstance(module_graph, weakref.ProxyTypes)
+        self.module_graph = module_graph
+        self.module_name = module_name
+        self.hook_filename = hook_filename
+
+        # Name of the in-memory module fabricated to refer to this hook script.
+        self.hook_module_name = (
+            hook_module_name_prefix + self.module_name.replace('.', '_'))
+
+        # Attributes defined by the _load_hook_module() method. To simplify this
+        # method, hook attributes are initialized to sane defaults.
+        self._hook_module = None
+        self._hook_attrs = {
+            'datas': set(),
+            'binaries': set(),
+            'excludedimports': set(),
+            'hiddenimports': list(),
+        }
+
+
+    def __getattr__(self, attr_name):
+        '''
+        Get the magic attribute with the passed name (e.g., `datas`) from this
+        lazily loaded hook script if any _or_ raise `AttributeError` otherwise.
+
+        This special method is called only for attributes _not_ already defined
+        by this object. This includes both magic and undefined attributes.
+
+        See Also
+        ----------
+        Class docstring for supported magic attributes.
+        '''
+
+        # If this is a magic attribute, initialize this attribute by lazy
+        # loading this hook script and then return this attribute.
+        if attr_name in _MAGIC_MODULE_HOOK_ATTR_NAMES:
+            self._load_hook_module()
+            return self._hook_attrs[attr_name]
+        # Else, this is an undefined attribute. Raise an exception.
+        else:
+            raise AttributeError(attr_name)
+
+
+    def __setattr__(self, attr_name, attr_value):
+        '''
+        Set the attribute with the passed name to the passed value.
+
+        If this is a magic attribute, this hook script will be lazily loaded
+        before setting this attribute. Unlike `__getattr__()`, this special
+        method is called to set _any_ attribute -- including magic, non-magic,
+        and undefined attributes.
+
+        See Also
+        ----------
+        Class docstring for supported magic attributes.
+        '''
+
+        # If this is a magic attribute, set this attribute to the passed value.
+        if attr_name in _MAGIC_MODULE_HOOK_ATTR_NAMES:
+            # Validate this attribute's new value.
+            if attr_name in {'datas', 'binaries', 'excludedimports'}:
```

Comment:
> IMHO this is over-engineered.

That's valid criticism. You're probably right. Synchronizing this validation with magic attribute names and types declared elsewhere is a recipe for death soup. And I'm tired of death soup.

So:

1. I'll remove this validation.
1. I'll add magic attribute types to `MAGIC_MODULE_HOOK_ATTR_NAMES` – probably by refactoring that `set` into a `dict`.